File: src/3-longest-substring-without-repeating-characters/solution.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
  char_set = Set.new
  res = l = 0
  s.chars.each_with_index do |v, r|
    while char_set.include? v
      char_set.delete(s[l])
      l += 1
    end
    char_set << v
    res = [res, r - l + 1].max
  end

  res
end

File: src/3-longest-substring-without-repeating-characters/solution.ts
function lengthOfLongestSubstring(s: string): number {
  let charSet = new Set<string>();
  let res: number = 0;
  let l: number = 0, r: number = 1;
  for(let char of s) {
    while(charSet.has(char)){
      charSet.delete(s[l]);
      l++;
    }
    charSet.add(char);
    res = Math.max(res, r - l);
    r++;
  }

  return res;
};
